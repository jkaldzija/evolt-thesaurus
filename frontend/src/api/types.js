export type Word = {
    synonyms: Array<string>,
    value: string
}
