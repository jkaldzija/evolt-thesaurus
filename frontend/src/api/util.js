import Axios, {AxiosInstance} from 'axios'
import {API_URL} from "../config";

export const apiClient: AxiosInstance = Axios.create({
    baseURL: API_URL
});
