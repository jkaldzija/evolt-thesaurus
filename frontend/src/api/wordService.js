import {apiClient} from "./util";
import {AxiosPromise, AxiosResponse} from "axios";
import type {Word} from "./types";

const WORD_PATH = "word";
const WORD_LIST_PATH = "word/list";


export default {
    getSynonymsForWord: (value: string): AxiosPromise<AxiosResponse<Word>> => apiClient.get(WORD_PATH, {
        params: {
            value
        }
    }),
    createWord: (word: Word): AxiosPromise<AxiosResponse<string>> => apiClient.post(WORD_PATH, word),
    listWordsPaginated: (offset: number, limit: number): AxiosPromise<AxiosResponse<Array<Word>>> => apiClient.get(WORD_LIST_PATH, {
        params: {
            offset,
            limit
        }
    })
}
