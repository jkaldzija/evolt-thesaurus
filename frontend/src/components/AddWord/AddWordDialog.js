import React from 'react';
import {Col, Input, Modal, ModalBody, ModalHeader, Row} from "reactstrap";
import wordService from "../../api/wordService";
import {toast} from "react-toastify";
import style from "./style.module.css";
import SynonymFormItem from "./SynonymFormItem";
import ThemeButton from "../Form/ThemeButton/ThemeButton";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus, faTimes } from '@fortawesome/free-solid-svg-icons';
import * as classnames from "classnames";
import {notifyError, notifySuccess} from "../Utility/notification";


// Set prop types
type AddSynonymDialogProps = {
    isOpen: boolean,
    toggle: () => void,
};

class AddWordDialog extends React.Component<AddSynonymDialogProps> {
    state = {
        value: "",
        synonyms: [""]
    };

    addSynonym = () => this.setState({
        synonyms: [...this.state.synonyms, ""]
    });

    removeSynonym = index => () => {
        const synonyms = [...this.state.synonyms];
        synonyms.splice(index, 1);
        this.setState({synonyms})
    };

    handleWordChange = event => this.setState({
        value: event.target.value
    });

    handleSynonymChange = index => value => {
        const synonyms = [...this.state.synonyms];
        synonyms[index] = value;
        this.setState({synonyms})
    };

    onClose = () => {
        this.setState({
            value: "",
            synonyms: []
        });
        this.props.toggle();
    };

    onSave = async (e) => {
        e.preventDefault();
        try {
            await wordService.createWord(this.state);
            notifySuccess("Word saved");
            this.onClose();
        } catch (e) {
            notifyError("Error happened while saving");
            console.log(e);
        }
    };

    renderSynonyms = () => {
        return this.state.synonyms.map((synonym, index) => {
            return <SynonymFormItem key={`${index}`} value={synonym} onChange={this.handleSynonymChange(index)} onDelete={this.removeSynonym(index)}/>
        })
    };

    render() {
        const {isOpen} = this.props;
        const {value, synonyms} = this.state;
        return (
            <Modal isOpen={isOpen} size={"lg"} toggle={() => this.onClose()}>
                    <ModalHeader className={classnames("typography-bold", style.header)}>

                        <div className={"d-flex justify-content-between w-100"}>
                            <div>Add word</div>
                            <FontAwesomeIcon className={style.close} onClick={() => this.onClose()} icon={faTimes}/>
                        </div>

                    </ModalHeader>

                <ModalBody>
                    <form onSubmit={(e) => this.onSave(e)}>
                        <Row>
                            <Col sm={12} md={{size:6}} className="mb-3">
                                <label htmlFor={"word"}>Word</label>
                                <Input label="Word"
                                       className={style.input}
                                       value={value}
                                       name={"word"}
                                       placeholder={"Type the word"}
                                       onChange={this.handleWordChange}/>
                            </Col>
                            <Col sm={12} md={{ size: 12, offset: 0 }} >
                                <label>Synonyms</label>
                                {this.renderSynonyms()}

                            </Col>
                            <Col sm={12} >
                                <ThemeButton onClick={(e) => this.addSynonym()}>
                                    <FontAwesomeIcon icon={faPlus} className={"mr-2"} />
                                    Add synonym
                                </ThemeButton>
                            </Col>
                            <Col sm={12} md={{size: 6, offset: 6}}>
                                <div className="d-flex justify-content-end">
                                    <ThemeButton className={style.submit}>
                                        Submit
                                    </ThemeButton>
                                </div>
                            </Col>
                        </Row>
                    </form>
                </ModalBody>
            </Modal>
        );
    };

}

// Set default props
AddWordDialog.defaultProps = {
    open: false,
    onClose: () => {},
};

export default AddWordDialog;
