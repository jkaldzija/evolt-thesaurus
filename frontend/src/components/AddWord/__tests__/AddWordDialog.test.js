import React from 'react';
import {shallow} from "enzyme";
import {apiClient} from "../../../api/util";
import AddWordDialog from "../AddWordDialog";
import MockAdapter from 'axios-mock-adapter';

describe('Add Word Dialog', function () {
    it('should render correctly', function () {
        const component = shallow(
            <AddWordDialog open={true} onClose={() => {}}/>
        );
        component.setState({
            value: "test",
            synonyms: [
                "syn1",
                "syn2",
                "syn3"
            ]
        });
        expect(component).toMatchSnapshot();
    });
    it('should handle input change', function () {
        const component = shallow(
            <AddWordDialog open={true} onClose={() => {}}/>
        );
        component.instance().handleWordChange({target: {value: "test"}});
        expect(component.state("value")).toBe("test");
    });
    it('should handle adding synonyms', function () {
        const component = shallow(
            <AddWordDialog open={true} onClose={() => {}}/>
        );
        component.instance().addSynonym();
        expect(component.state("synonyms").length).toBe(1);
    });
    it('should handle removing synonyms', function () {
        const component = shallow(
            <AddWordDialog open={true} onClose={() => {}}/>
        );
        component.setState({
            value: "test",
            synonyms: [
                "syn1",
                "syn2",
                "syn3"
            ]
        });
        component.instance().removeSynonym(1)();
        expect(component.state("synonyms").length).toBe(2);
    });
    it('should handle synonym input', function () {
        const component = shallow(
            <AddWordDialog open={true} onClose={() => {}}/>
        );
        component.setState({
            value: "test",
            synonyms: [
                "syn1",
                "syn2",
                "syn3"
            ]
        });
        component.instance().handleSynonymChange(1)("test");
        expect(component.state("synonyms")[1]).toBe("test");
    });
    it('should reset state on close', function () {
        const component = shallow(
            <AddWordDialog open={true} onClose={() => {}}/>
        );
        component.setState({
            value: "test",
            synonyms: [
                "syn1",
                "syn2",
                "syn3"
            ]
        });
        component.instance().onClose();
        expect(component.state()).toEqual({
            value: "",
            synonyms: []
        });
    });
    it('should handle sending new word request', async function () {
        const wordData = {
            value: "test",
            synonyms: [
                "syn1",
                "syn2",
                "syn3"
            ]
        };
        const closeCallback = jest.fn();
        const component = shallow(
            <AddWordDialog open={true} onClose={closeCallback}/>
        );
        component.setState(wordData);
        const mock = new MockAdapter(apiClient);
        mock.onPost('word', wordData).reply(200, "success");
        await component.instance().onSave();
        expect(closeCallback).toHaveBeenCalled()
    });
    it('should handle word request errors', async function () {
        const wordData = {
            value: "test",
            synonyms: [
                "syn1",
                "syn2",
                "syn3"
            ]
        };
        const closeCallback = jest.fn();
        const component = shallow(
            <AddWordDialog open={true} onClose={closeCallback}/>
        );
        component.setState(wordData);
        const mock = new MockAdapter(apiClient);
        mock.onPost('word', wordData).reply(400, "error");
        await component.instance().onSave();
        expect(closeCallback).toHaveBeenCalledTimes(0)
    });
});
