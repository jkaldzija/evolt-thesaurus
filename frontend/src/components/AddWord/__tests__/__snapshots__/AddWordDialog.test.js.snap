// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Word Dialog should render correctly 1`] = `
<WithStyles(Dialog)
  aria-labelledby="add-dialog-title"
  fullWidth={true}
  onClose={[Function]}
  open={true}
>
  <WithStyles(DialogTitle)
    id="add-dialog-title"
  >
    Add new word
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <TextField
      fullWidth={true}
      label="Word"
      onChange={[Function]}
      required={false}
      select={false}
      value="test"
      variant="standard"
    />
    <div
      className="add-synonym-list-container"
    >
      <WithStyles(Typography)
        variant="subheading"
      >
        Synonyms:
      </WithStyles(Typography)>
      <SynonymFormItem
        key="0"
        onChange={[Function]}
        onDelete={[Function]}
        value="syn1"
      />
      <SynonymFormItem
        key="1"
        onChange={[Function]}
        onDelete={[Function]}
        value="syn2"
      />
      <SynonymFormItem
        key="2"
        onChange={[Function]}
        onDelete={[Function]}
        value="syn3"
      />
      <div
        className="synonym-add-button-container"
      >
        <WithStyles(Button)
          color="primary"
          onClick={[Function]}
        >
          <pure(AddIcon) />
           Add synonym
        </WithStyles(Button)>
      </div>
    </div>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Save word
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;
